aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -45,41 +45,33 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
   return r;
 }
 
-Tensor to(const Tensor& self, const TensorOptions& options_, bool non_blocking, bool copy, c10::optional<c10::MemoryFormat> optional_memory_format) {
-
+Tensor to(
+  const Tensor& self,
+  const TensorOptions& options_,
+  bool non_blocking,
+  bool copy,
+  c10::optional<c10::MemoryFormat> optional_memory_format
+) {
   TORCH_CHECK(
     !(options_.has_memory_format() && optional_memory_format.has_value()),
     "Cannot set memory_format both in TensorOptions and explicit argument; please delete "
     "the redundant setter.");
   auto options = options_.merge_in(TensorOptions().memory_format(optional_memory_format));
 
-  auto memory_format = options.memory_format_opt().value_or(MemoryFormat::Contiguous);
-
   TORCH_CHECK(options.requires_grad_opt() == c10::nullopt,
            "to(options) expects unset requires_grad flag, but got "
            "options.requires_grad set as ", options.requires_grad());
 
-  const auto & layout_opt = options.layout_opt();
-  TORCH_CHECK(!layout_opt || self.layout() == layout_opt.value(),
+  TORCH_CHECK(!options.has_layout() || self.layout() == options.layout(),
            "to(options) doesn't support converting to a different layout, "
            "but got self.layout being ", self.layout(),
            " and options.layout set as ", options.layout());
 
-  // TODO: refactor all of this code to just use merge_in
-
-  auto device_opt = options.device_opt();
-  if (device_opt) {
-    device_opt = ensure_has_index(device_opt.value());
-  }
-  const auto & dtype_opt = options.dtype_opt();
-  auto specified_options = self.options();
-  if (device_opt) {
-    specified_options = specified_options.device(device_opt.value());
-  }
-  if (dtype_opt) {
-    specified_options = specified_options.dtype(dtype_opt.value());
+  if (options.has_device()) {
+    options = options.device(ensure_has_index(options.device()));
```

Comment:
this is kind of smelly but no need to handle it in this pr

Reply:
Yeah it's definitely a wrinkle in the carpet. Not sure what would be better if we want to use `merge_in`, which we definitely do... doing the ensure on the merged options seems marginally worse.

Are you thinking of something nonlocal that would spare us the having to ensure at all?